src/t_hash.c

```diff
@@ -29,52 +29,183 @@
-    size_t total_size = sizeof(hashTypeEntry) + field_size;
-    hashTypeEntry *entry = zmalloc(total_size);
-
-    entry->value = value;
-    entry->field_offset = sdsHdrSize(field_sds_type);
-    sdswrite(entry->field_data, field_size, field_sds_type, field, field_len);
-    return entry;
+    size_t value_len = sdslen(value);
+    size_t value_size = sdsReqSize(value_len, SDS_TYPE_8);
+    sds embedded_field_sds;
+    if (field_size + value_size <= EMBED_VALUE_MAX_ALLOC_SIZE) {
+        /* Embed field and value. Value is fixed to SDS_TYPE_8.
+         *
+         *     +--------------+---------------+
+         *     | field        | value         |
+         *     | hdr "foo" \0 | hdr8 "bar" \0 |
+         *     +--------------+---------------+
+         */
+        size_t min_size = field_size + value_size;
+        size_t buf_size;
+        char *buf = zmalloc_usable(min_size, &buf_size);
+        embedded_field_sds = sdswrite(buf, field_size, field_sds_type, field, field_len);
+        sdswrite(buf + field_size, buf_size - field_size, SDS_TYPE_8, value, value_len);
+        /* Field sds aux bits are zero, which we use for this entry encoding. */
+        sdsSetAuxBit(embedded_field_sds, FIELD_SDS_AUX_BIT_ENTRY_HAS_VALUE_PTR, 0);
+        serverAssert(!entryHasValuePtr(embedded_field_sds));
+        sdsfree(value);
+    } else {
+        /* Embed field, but not value. Field must be >= SDS_TYPE_8 to encode to
+         * indicate this type of entry.
+         *
+         *     +-------+---------------+
+         *     | value | field         |
+         *     | ptr   | hdr8 "foo" \0 |
+         *     +-------+---------------+
+         */
+        char field_sds_type = sdsReqType(field_len);
+        if (field_sds_type == SDS_TYPE_5) field_sds_type = SDS_TYPE_8;
+        field_size = sdsReqSize(field_len, field_sds_type);
+        size_t alloc_size = sizeof(sds *) + field_size;
+        char *buf = zmalloc(alloc_size);
+        *(sds *)buf = value;
+        embedded_field_sds = sdswrite(buf + sizeof(sds *), field_size, field_sds_type, field, field_len);
+        /* Store the entry encoding type in sds aux bits. */
+        sdsSetAuxBit(embedded_field_sds, FIELD_SDS_AUX_BIT_ENTRY_HAS_VALUE_PTR, 1);
+        serverAssert(entryHasValuePtr(embedded_field_sds));
+    }
+    return (void *)embedded_field_sds;
 }
 
+/* The entry pointer is the field sds, but that's an implementation detail. */
 sds hashTypeEntryGetField(const hashTypeEntry *entry) {
-    const char *field = entry->field_data + entry->field_offset;
-    return (sds)field;
+    return (sds)entry;
 }
 
 sds hashTypeEntryGetValue(const hashTypeEntry *entry) {
-    return entry->value;
+    if (entryHasValuePtr(entry)) {
+        return *hashTypeEntryGetValueRef(entry);
+    } else {
+        /* Skip field content, field null terminator and value sds8 hdr. */
+        size_t offset = sdslen(entry) + 1 + sdsHdrSize(SDS_TYPE_8);
+        char *buf = (char *)entry + offset;
```

Comment:
nit: I cannot explain why it bothers me :) 
```suggestion
        sds buf = (char *)entry + offset;
```

Reply:
You need to explain. Otherwise, I don't take your comment as valid. :laughing:

An `sds` variable named `buf` would bother me. `char *buf` is used when working with the underlying bytes. When working with the sds string abstractions, I'd call it `sds value` or something.

I'm fine with `return (char *)entry + offset;` though, or if you insist `return (sds)((char *)entry + offset);`

Reply:
I think for the same reason you choose to cast char* to sds* via void*. there is no real type change only alias change right? It just seem like we should return the same "type" even though there is not special need to since sds is an alias to char*

Reply:
but as I said. not important...

Reply:
Here you're right that `sds = char *` so there's not really a cast. I think the intuition and the naming is somewhat relevant here though. Your suggestion is good but I'll change `char *buf` to `sds value`. Or possibly to make it easier to follow why we get this offset, it can be done in multiple steps...

```C
        char *buf = entry;
        buf += sdslen(entry) + 1;      /* Move to after the embedded field */
        buf += sdsHdrSize(SDS_TYPE_8); /* Skip value header */
        return buf;
```

Or for brevity I'll stick with a single line, since there are some drawings and comments about this elsewhere.

```C
        return (char *)entry + offset;
```